### PR TITLE
Fix sliding window attention used in Gemma2FlashAttention2

### PR DESCRIPTION
**Pull request by brcps12**
# What does this PR do?

Currently, implementation of the sliding window in the `Gemma2FlashAttention2` module has issues. Specifically, when applying the sliding window, [the attention mask is sliced](https://github.com/huggingface/transformers/blob/f5cdbf6e546784639c2d8f72420198138d2a2a13/src/transformers/models/gemma2/modeling_gemma2.py#L575). This can lead to problems if the sequence length exceeds the sliding window size.

Instead, I use the `window_size` parameter supported by Flash Attention 2. This ensures the sliding window is applied correctly, regardless of sequence length.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @ArthurZucker

File: src/transformers/models/gemma2/modeling_gemma2.py [hunk 1]
```diff
@@ -569,18 +570,16 @@ def forward(
...
-            if self.config._attn_implementation == "flash_attention_2":
-                attention_mask = attention_mask[:, -self.sliding_window :]
-            else:
-                min_dtype = torch.finfo(hidden_states.dtype).min
-                sliding_window_mask = torch.tril(
-                    torch.ones_like(attention_mask, dtype=torch.bool), diagonal=-self.sliding_window
-                )
-                attention_mask = torch.where(sliding_window_mask, min_dtype, attention_mask)
-                if attention_mask.shape[-1] <= 1:  # when decoding
-                    attention_mask = attention_mask[:, :, :, -self.sliding_window :]
```
Comment (ArthurZucker) [inline src/transformers/models/gemma2/modeling_gemma2.py:583]: Hey! You mention that:
>  This can lead to problems if the sequence length exceeds the sliding window size.

But we are only doing this when decoding!

Comment (brcps12) [reply src/transformers/models/gemma2/modeling_gemma2.py:583]: The problematic areas are at [L575](https://github.com/huggingface/transformers/blob/7ad784ae9da9b8ce61ba734199fb258d8d95460f/src/transformers/models/gemma2/modeling_gemma2.py#L575) and [L385-L387](https://github.com/huggingface/transformers/blob/7ad784ae9da9b8ce61ba734199fb258d8d95460f/src/transformers/models/gemma2/modeling_gemma2.py#L385-L387). The issue arises with the following parameters and inputs:

```python
sliding_window = 4
# shape [2, 8]
input_ids = [
    [1, 2, 3, 4, 5, 6, 0, 0],
    [7, 8, 9, 10, 0, 0, 0, 0],
]
attention_mask = [
    [1, 1, 1, 1, 1, 1, 0, 0],
    [1, 1, 1, 1, 0, 0, 0, 0],
]
```
In this example, the `attention_mask` values are sliced to the last 4 values, `[[1, 1, 0, 0], [0, 0, 0, 0]]`, and the `key` and `value` are derived from the first 4 tokens of input_ids, `[[1, 2, 3, 4], [7, 8, 9, 10]]`. Therefore, the mask `[0, 0, 0, 0]` is passed to the flash attention function, it becomes impossible to compute proper attention. This issue occurs not only during decoding but also during training, regardless of whether decoding is involved.

Additionally, I have a question: when you mention “decoding,” are you referring to the `model.generate` function?

Comment (ArthurZucker) [reply src/transformers/models/gemma2/modeling_gemma2.py:583]: Sorry, yeah the slicing should only happen when you are indeed decoding, decoding means generating yep. 
The first forward pass should not slice the mask

**Review by ArthurZucker — COMMENTED** [review]
Good catch, can you please commit something with tag `[run-slow] gemma` to run the slow tests?

**Comment by brcps12** [comment]
I committed with the message.

Additionally, regardless of whether the tests pass, an error occurs when generating sentences with the current implementation of `gemma2 + flash_attn2`. That likely needs to be fixed

I think it could be implemented by referring to the [MistralFlashAttention2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/mistral/modeling_mistral.py#L320-L347) code.

Should I make these changes in this PR?

**Review by ArthurZucker — COMMENTED** [review]
Can you resolve the conflicts? 🤗

**Comment by brcps12** [comment]
Done.

I added a condition `past_key_value is not None` for decoding phase.

File: src/transformers/models/gemma2/modeling_gemma2.py [hunk 2]
```diff
@@ -567,7 +568,8 @@ def forward(
         if self.is_sliding and attention_mask is not None:  # efficient SDPA and no padding
             # Flash-attn is a 2D tensor
             if self.config._attn_implementation == "flash_attention_2":
-                attention_mask = attention_mask[:, -self.sliding_window :]
+                if past_key_value is not None:  # when decoding
+                    attention_mask = attention_mask[:, -self.sliding_window :]
```
Comment (ArthurZucker) [inline src/transformers/models/gemma2/modeling_gemma2.py:572]: Ah that looks better yeah, when you have the first forward pass you nee the attention mask, full but have to use `self.sliding_window` when calling attention, but then when you decode the mask is too big, and the kv cache is sliding_window -1

**Review by ArthurZucker — APPROVED** [review]
Looks good now!

**Comment by ArthurZucker** [comment]
Thanks @brcps12

**Comment by kiddj** [comment]
It seems that @brcps12 submitted this PR, but his name was unfortunately not mentioned in the release notes. Could we update the release notes to include his contribution?

**Comment by ArthurZucker** [comment]
yeah for sure

**Comment by ArthurZucker** [comment]
Sorry about that
